### PR TITLE
Heterogeneity analysis feature for pre-computed predictions

### PR DESCRIPTION
**Pull request by DonggyuBan**
Relates to #53

# Description

This pull request introduces a heterogeneity detection feature to identify bias in a model's binary prediction outcomes across demographic groups. The feature analyzes a CSV file of pre-computed predictions to detect heterogeneity by group using accuracy and AUC.

```heterogeneity_report.py```: loads a CSV file with a configurable group column, calculates per-group accuracy and AUC, and uses statistical tests to find heterogeneity.

```heterogeneity_eval.py```: converts the CSV file into the Inspect AI format and returns standardized metrics from the analysis.

## New Dependencies

(None)

# Testing Instructions

Run ```heterogeneity_eval.py``` with pre-computed CSV file. 
Use: 
```inspect eval heterogeneity_eval.py -- --input_file file_name.csv --group_column GROUP_NAME```

### CODE REVIEW

File: heterogeneity/heterogeneity_eval.py [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+
+from heterogeneity_report import (
+    calculate_group_metrics,
+    find_heterogeneity,
+    identify_groups,
+    visualizations,
+    generate_report
+)
+
+accumulated_data = []
```
Comment (niznik-dev) [inline heterogeneity/heterogeneity_eval.py:15]: This is not ideal - I think we can reconsider how we handle this list.

For example, can we define it in a task and then pass it to the relevant functions instead? I tend to reserve globals for constants.

File: heterogeneity/heterogeneity_eval.py [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+EXPECTED_SAMPLES = None
```
Comment (niznik-dev) [inline heterogeneity/heterogeneity_eval.py:16]: Similarly, using all caps for this variable name implies it is a constant, but I see it gets changed below. This too is not ideal.

Comment (niznik-dev) [reply heterogeneity/heterogeneity_eval.py:16]: But I'll let you respond because it may be we're locked into this based on how inspect_ai works.

File: heterogeneity/heterogeneity_eval.py [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+            return metrics_dict
+        
+        else:
+            return {'samples_processed': float(total_samples)}
+    
+    return compute
+
+@scorer(metrics=[heterogeneity_metric()])
+def heterogeneity_scorer():
+
```
Comment (niznik-dev) [inline heterogeneity/heterogeneity_eval.py:124]: I would remove this empty line

File: heterogeneity/heterogeneity_eval.py [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+        else:
+            return {'samples_processed': float(total_samples)}
+    
+    return compute
+
+@scorer(metrics=[heterogeneity_metric()])
+def heterogeneity_scorer():
+    
+    async def score(state, target, **kwargs):
+        if hasattr(target, 'target'):
```
Comment (niznik-dev) [inline heterogeneity/heterogeneity_eval.py:126]: That there are three possibilities here is quite confusing - can you explain what's going on? We may be able to simplify this.

File: heterogeneity/heterogeneity_eval.py [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+    print(f"Loading data from {csv_path}...")
+    df = pd.read_csv(csv_path)
+    
+    EXPECTED_SAMPLES = len(df)
+    
+    print(f"Loaded {len(df)} rows.")
+    print(f"Using group column: {group_column}")
+    
+    unique_groups = df[group_column].unique()
+    n_groups = len(unique_groups)
```
Comment (niznik-dev) [inline heterogeneity/heterogeneity_eval.py:31]: [minor] I would probably call this `n_unique_groups`

File: heterogeneity/heterogeneity_eval.py [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+
+@metric
+def heterogeneity_metric():
+    def compute(scores):
+        global accumulated_data, EXPECTED_SAMPLES
+        
+        new_scores = scores[len(accumulated_data):]
+        
+        for score in new_scores:
+            if score.metadata:
```
Comment (niznik-dev) [inline heterogeneity/heterogeneity_eval.py:66]: When would a score not have metadata?

File: heterogeneity/heterogeneity_eval.py [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+                'group': state.metadata.get("group")
+            }
+        )
+    
+    return score
+
+@task
+def heterogeneity_task():
+
+    input_file = os.environ.get('INPUT_FILE', 'predictions.csv')
```
Comment (niznik-dev) [inline heterogeneity/heterogeneity_eval.py:147]: I'm a bit skeptical that these should be env variables but we can discuss

File: heterogeneity/heterogeneity_report.py [hunk 8]
```diff
@@ -0,0 +1,248 @@
+import pandas as pd
```
Comment (niznik-dev) [inline heterogeneity/heterogeneity_report.py:1]: Same comment on import organization

File: heterogeneity/heterogeneity_report.py [hunk 9]
```diff
@@ -0,0 +1,248 @@
...
+import argparse
+import os
+from sklearn.metrics import accuracy_score, roc_auc_score 
+from scipy.stats import f_oneway
+
+def load_data(file_path, group_column='GROUP'):
+    """Load CSV data for analysis"""
+    df = pd.read_csv(file_path)
+    
+    if group_column != 'GROUP' and group_column in df.columns:
```
Comment (niznik-dev) [inline heterogeneity/heterogeneity_report.py:14]: What happens if it's not 'GROUP' and `group_column` isn't in `df.columns`?

File: heterogeneity/heterogeneity_report.py [hunk 10]
```diff
@@ -0,0 +1,248 @@
...
+        df['GROUP'] = df[group_column]
```
Comment (niznik-dev) [inline heterogeneity/heterogeneity_report.py:15]: [minor] This isn't the most memory efficient way to accomplish this, but I think it's fine for our purposes

File: heterogeneity/heterogeneity_report.py [hunk 11]
```diff
@@ -0,0 +1,248 @@
...
+    
+    for group in df['GROUP'].unique():
+        group_df = df[df['GROUP'] == group]
+        
+        accuracy = accuracy_score(
+            group_df['TRUE_LABEL'],
+            group_df['PREDICTION']
+        )
+        
+        try:
```
Comment (niznik-dev) [inline heterogeneity/heterogeneity_report.py:32]: Oh I'm so glad to see try/except and not assert!

File: heterogeneity/heterogeneity_report.py [hunk 12]
```diff
@@ -0,0 +1,248 @@
...
+                    'z_score': 0
+                })
+                auc_outlier = True
+        
+        if acc_outlier and auc_outlier:
+            identified_groups['outlying_in_both'].append(group)
+    
+    return identified_groups
+
+def visualizations(group_metrics, output_dir='results'):
```
Comment (niznik-dev) [inline heterogeneity/heterogeneity_report.py:147]: I knew there was something I didn't like about this when I saw it called in `run_analysis()`, and now I know what it is. This should be a verb so it reads better. Maybe `visualize_data`?

Comment (niznik-dev) [reply heterogeneity/heterogeneity_report.py:147]: Or `create_performance_visualizations` as is below could work - up to you

File: heterogeneity/heterogeneity_report.py [hunk 13]
```diff
@@ -0,0 +1,248 @@
...
+            'heterogeneity_found': bool(heterogeneity_results['accuracy_heterogeneity']['heterogeneity'] or 
+                                 heterogeneity_results['auc_heterogeneity']['heterogeneity']),
+        },
+        'heterogeneity_analysis': heterogeneity_results,
+        'identified_groups': identified_groups,
+        'group_metrics': group_metrics_json
+    }
+    
+    with open(f'{output_dir}/heterogeneity_report.json', 'w') as f:
+        json.dump(report, f, indent=2, default=str)
```
Comment (niznik-dev) [inline heterogeneity/heterogeneity_report.py:219]: Can you double-check the order this prints the values? I know I've needed to disable auto-alphabetizing behavior before when doing `json.dump` to make things a bit more readable.

File: heterogeneity/heterogeneity_report.py [hunk 14]
```diff
@@ -0,0 +1,248 @@
...
+    
+    group_metrics = calculate_group_metrics(df)
+    heterogeneity_results = find_heterogeneity(df, group_metrics)
+    identified_groups = identify_groups(group_metrics)
+    
+    visualizations(group_metrics, output_dir)
+    report = generate_report(df, group_metrics, heterogeneity_results, identified_groups, output_dir)
+    
+    print(f"\nAnalysis complete. Results saved to {output_dir}/")
+    return report
```
Comment (niznik-dev) [inline heterogeneity/heterogeneity_report.py:239]: This is superfluous, right? You shouldn't need to return anything since it's unused.

**Review by niznik-dev — CHANGES_REQUESTED** [review]
I think this is heading in the right direction. Some topics to discuss though before I'd approve it. Well done in general (from what I can follow!).

File: heterogeneity/heterogeneity_eval.py (hunk 2 above)
Comment (DonggyuBan) [reply heterogeneity/heterogeneity_eval.py]: Seems like the constraint with inspect_ai is that its metrics only accept a scores parameter. I tried using a class _MetricState to organize things better. It still uses global state, but seems cleaner than just global variables.

File: heterogeneity/heterogeneity_eval.py (hunk 6 above)
Comment (DonggyuBan) [reply heterogeneity/heterogeneity_eval.py]: Fixed it

File: heterogeneity/heterogeneity_report.py (hunk 9 above)
Comment (DonggyuBan) [reply heterogeneity/heterogeneity_report.py]: Added error handling code!

File: heterogeneity/heterogeneity_eval.py (hunk 7 above)
Comment (DonggyuBan) [reply heterogeneity/heterogeneity_eval.py:149]: I tried using argparse, but Inspect AI doesn't seem to allow it. Should I edit the script to use a configuration file instead?

File: heterogeneity/heterogeneity_eval.py [hunk 15]
```diff
@@ -1,40 +1,47 @@
...
     find_heterogeneity,
     identify_groups,
-    visualizations,
+    visualize_data,
     generate_report
 )
 
-accumulated_data = []
-EXPECTED_SAMPLES = None
+class _MetricState:
```
Comment (niznik-dev) [inline heterogeneity/heterogeneity_eval.py:15]: I think I can get behind this, good idea

**Comment by niznik-dev** [comment]
I think we're just down to very minor import optimization and a config file vs environment variables. Well, and probably @alessandramaranca approving as well.
